apis/src/pages/register.rs

```diff
@@ -1,68 +1,203 @@
+use crate::functions;
+use crate::functions::users::get::UsernameTaken;
 use crate::{components::organisms::header::Redirect, providers::auth_context::AuthContext};
+use lazy_static::lazy_static;
+use leptos::leptos_dom::helpers::debounce;
 use leptos::*;
 use leptos_router::ActionForm;
+use leptos_use::use_debounce_fn_with_arg;
+use regex::Regex;
+use std::time::Duration;
+use web_sys::Event;
+
+const BANNED_USERNAMES: [&str; 3] = ["black", "white", "admin"];
+const VALID_USERNAME_CHARS: &str = "-_";
+
+lazy_static! {
+    static ref EMAIL_RE: Regex =
+        Regex::new(r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$").unwrap();
+}
 
 #[component]
 pub fn Register(#[prop(optional)] extend_tw_classes: &'static str) -> impl IntoView {
     let auth_context = expect_context::<AuthContext>();
+    let username_taken = create_server_action::<UsernameTaken>();
     let pathname =
         move || use_context::<Redirect>().unwrap_or(Redirect(RwSignal::new(String::from("/"))));
     let my_input = NodeRef::<html::Input>::new();
     create_effect(move |_| {
         let _ = my_input.get_untracked().map(|el| el.focus());
     });
+    let username = RwSignal::new(String::new());
+    let email = RwSignal::new(String::new());
+    let pw = RwSignal::new(String::new());
+    let pw_confirm = RwSignal::new(String::new());
+    let has_invalid_char = RwSignal::new(false);
+    let is_invalid_email = RwSignal::new(false);
+    let throttled_input = move |signal_to_update| {
+        use_debounce_fn_with_arg(update_from_input(signal_to_update), 350.0)
+    };
 
+    let pw_invalid = move || {
+        let pw = pw();
+        !(7 < pw.len() && pw == pw_confirm())
```

Comment:
We should check here that it's not greater than 128 chars :)

Reply:
the form prevents you from going over, isn't that enough?